### PR TITLE
Compilation and runtime stability improvement

### PR DESCRIPTION
**Pull request by prclibo**
Hi, was trying to compile camodocal on a old centos server and did some changes:

- Removed custom FindOpenCV s.t. the offical OpenCVConfig.cmake is used. 
- Removed custom CMAKE_LIBRARY_PATH and CMAKE_INCLUDE_PATH. They messed up default cmake search order for custom install libraries.
- Some std constainers of Eigen object forgot to declare allocator.
- Some runtime bug fixes.

 @ahundt @hengli  Could you please take a look if these fixes work on your side?

Thanks!

### CODE REVIEW

**Comment by ahundt** [comment]
Did you check your cmake opencv changes on both 2.x and 3.x? Your changes sound like something I hoped to do but didn't get around to, so that's great!

**Comment by ahundt** [comment]
I also think your switch from unordered map to std map will harm performance unless there is some other compelling reason you found during testing

**Comment by prclibo** [comment]
@ahundt 
1) I think it works for opencv 2 and 3 on my side and currently I am using opencv 3. It would be nice if you can test it on your side.
2) Yes map is slow. The solution is whether we find out the correct form to declare a boost::unordered_map of Eigen::Vector3d, which I failed to, or enable c++11.

**Comment by prclibo** [comment]
@ahundt fixed unordered_map. plz take a look:)

File: CMakeLists.txt [hunk 1]
```diff
@@ -146,7 +122,7 @@ if(CAMODOCAL_PLATFORM_UNIX)
     #set_linker_flags("-Wl,--no-undefined -ldl -lnsl -lm -pthread -lrt" CACHE)
     find_package(LibDL)
     find_package(Libnsl)
-    find_package(LibM)
+    # find_package(LibM)
```
Comment (ahundt) [inline CMakeLists.txt]: Is libM no longer required? Also we have to be careful about capitalization for case sensitive/unsensitive filesystems.

File: CMakeLists.txt [hunk 2]
```diff
@@ -191,6 +167,7 @@ include_directories(
   include
 )
 
+message("GLOG_FOUND in main" ${GLOG_FOUND})
```
Comment (ahundt) [inline CMakeLists.txt]: shouldn't add random cmake prints

File: cmake/LibraryConfig.cmake [hunk 3]
```diff
@@ -84,7 +84,7 @@ endif()
 
 ##### Boost #####
 # Expand the next statement if newer boost versions than 1.40.0 are released
-set(Boost_ADDITIONAL_VERSIONS "1.40" "1.40.0" "1.49" "1.49.0")
+# set(Boost_ADDITIONAL_VERSIONS "1.40" "1.40.0" "1.49" "1.49.0")
```
Comment (ahundt) [inline cmake/LibraryConfig.cmake]: commented for a particular reason?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -3,6 +3,7 @@ add_subdirectory(ceres-solver)
 endif()
 add_subdirectory(brisk)
 add_subdirectory(calib)
+message("GLOG in src" ${GLOG_FOUND})
```
Comment (ahundt) [inline src/CMakeLists.txt]: random cmake print

File: src/brisk/CMakeLists.txt [hunk 5]
```diff
@@ -11,7 +11,7 @@ include_directories(include)
 
 #build the brisk library dynamic and static versions
 camodocal_library(camodocal_brisk SHARED ${BRISK_SOURCE_FILES} ${BRISK_HEADER_FILES})
-camodocal_link_libraries(camodocal_brisk agast ${OPENCV_CORE_LIBRARY})
+camodocal_link_libraries(camodocal_brisk agast ${OpenCV_LIBS})
```
Comment (ahundt) [inline src/brisk/CMakeLists.txt]: this expands the number of things linked dramatically compared to before.... ditto for all other cases of this

File: src/calib/CamOdoCalibration.cc [hunk 6]
```diff
@@ -173,6 +173,11 @@ CamOdoCalibration::calibrate(Eigen::Matrix4d& H_cam_odo)
     tvecsOdo.resize(mSegments.size());
     rvecsCam.resize(mSegments.size());
     tvecsCam.resize(mSegments.size());
+    if (mSegments.empty())
```
Comment (ahundt) [inline src/calib/CamOdoCalibration.cc]: nice addition

File: src/calib/CameraRigBA.cc [hunk 7]
```diff
@@ -242,7 +242,8 @@ CameraRigBA::run(int beginStage, bool optimizeIntrinsics,
         // For each scene point, record its coordinates with respect to the
         // first camera it was observed in.
 
-        boost::unordered_map<Point3DFeature*, Eigen::Vector3d> scenePointMap;
+        boost::unordered_map<Point3DFeature*, Eigen::Vector3d, boost::hash<Point3DFeature*>, std::equal_to<Point3DFeature*>, Eigen::aligned_allocator<std::pair<Point3DFeature* const, Eigen::Vector3d> > > scenePointMap;
```
Comment (ahundt) [inline src/calib/CameraRigBA.cc]: looks good

File: src/camera_models/CMakeLists.txt [hunk 8]
```diff
@@ -2,7 +2,8 @@ include_directories(
   ${CERES_INCLUDE_DIRECTORIES}
 )
 
-if(GLOG_FOUND AND OPENCV_FOUND)
+message("CV "${OpenCV_FOUND} "CV2" ${OpenCV_CORE_FOUND} "G " ${GLOG_FOUND})
```
Comment (ahundt) [inline src/camera_models/CMakeLists.txt]: could you explain this one? perhaps another print that isn't needed?

File: src/camera_models/CMakeLists.txt [hunk 9]
```diff
@@ -15,9 +16,7 @@ camodocal_library(camodocal_camera_models SHARED
 
 camodocal_link_libraries(camodocal_camera_models
   ${CAMODOCAL_PLATFORM_UNIX_LIBRARIES}
-  ${OPENCV_CORE_LIBRARY}
-  ${OPENCV_IMGPROC_LIBRARY}
-  ${OPENCV_CALIB3D_LIBRARY}
+  ${OpenCV_LIBS}
```
Comment (ahundt) [inline src/camera_models/CMakeLists.txt]: more deps than necessary... can this be refined?

**Comment by ahundt** [comment]
Ok I was able to review things! Looks great overall.

Do the OpenCV config scripts allow differentiation between OpenCV with contrib and without it?

Also, I haven't actually gone through the process of using the whole thing or I would do this myself, but could you perhaps make some notes on setting up and using everything? I've only used the hand eye calib myself.

**Comment by prclibo** [comment]
@ahundt thanks for the patient check! I have removed these random prints.

Regarding the opencv cmake issues
The official OpenCVConfig provides variables like OPENCV_XFEATURES2D_FOUND=1 when contrib modules are found. Also OpenCV_LIBS will include the name of contributed modules. I could add some check if you wanna make sure that cuda sift is shipped.
The OpenCV_LIBS add unnecessary libs but I don' find suitable solution in the official OpenCVConfig. Also I suggest not worry about this as this does not add too much load.

Regarding running the whole thing:
The intrinsic calib can be run easily. The original readme should be enough. But I cannot provide data for extrinsic calibration yet. So maybe we just leave this for the future.

File: CMakeLists.txt (hunk 1 above)
Comment (prclibo) [inline CMakeLists.txt]: Good question. What's LibM and what is the difference between Libm and LibM? I have no idea. Libm break my compilation and removing it seems not produce any effects.

Comment (ahundt) [inline CMakeLists.txt]: I believe that libM change will produce problems on other platforms. What happens when find_package(Libm) is used instead? that should fix the apparent capitalization change issue

**Comment by ahundt** [comment]
looks like it is almost there... just the libm issue and dealing with opencv_contrib/xfeatures2d

is an OPENCV_XFEATURES2D_FOUND check actually used anywhere? I think that will only work on 2.x, correct? For 3.0 OpenCV contrib is also required by the library so we need to detect when it isn't there and skip building components that require it plus output a warning. 

Here is one thing that checks, not sure if that's the best option:
https://stackoverflow.com/questions/33880515/opencv-in-cmake-based-project-checking-for-contrib-modules

**Comment by prclibo** [comment]
I am testing OPENCV_XFEATURES2D_FOUND on opencv 3.x so this should not be a problem. I think the xfeature2d check does not affect this pr so Let's put this in another pr and finish this one first. 

Libm is OK and can be found via cmake/FindLibm.cmake. I don't find any FindLibM.cmake in either camodocal or cmake system path which is why LibM break my compilation.

**Comment by ahundt** [comment]
Ok cool I'll merge it now and we can fix outstanding issues others encounter as they come up. Thanks for the nice changes!

**Comment by ahundt** [comment]
@prclibo it seems like there aren't any checks for opencv_contrib in the updated cmake scripts you provided, it succeeds without error in that case

**Pull request merged**